[ES|QL] Use same adhoc dataviews for queries with the same index pattern

stratoula opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/168131

We are using adhoc dataviews in ES|QL (Discover, Lens, Dashboards) only for technical reasons as these applications are made to work with these dataviews. We don't use neither the default timefield nor field formatters nor runtime fields so regardless of the ES|QL query, if the index pattern matches we don't need to create another adhoc dataview.

This comes with great advantages:

- we don't need to create a new dataview if there is already in the cache (performance boost especially on a dashboard, fewer calls to the fields caps api!)
-  solves this annoying bug https://github.com/elastic/kibana/issues/168131 

<img width="1771" alt="image" src="https://github.com/elastic/kibana/assets/17003240/51916dbc-590d-4142-9f70-7e0d7df4a5b7">

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented on packages/kbn-esql-utils/src/utils/index.ts:
ℹ️ Right now it only has one utility but I am going to move more from es-query and lang-editor in a follow-up PR. We need a package for ES|QL and I want to gather everything in one package

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

stratoula commented:
/ci

jughosta reviewed (approved):
Discover changes LGTM 👍

davismcphee reviewed (dismissed):
Sorry folks, I'm going to block this one temporarily just until we've had a chance to discuss a bit more as a team at our dev sync tomorrow.

dej611 reviewed (approved):
Code review only 👍

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
/ci

mattkime commented on packages/kbn-esql-utils/src/utils/index.ts:
```diff
@@ -0,0 +1,23 @@
+) {
+  return await dataViewsService.create({
+    title: indexPattern,
+    id: `esql-${indexPattern}`,
```
We need to see if there's a max size for the id field. I've seen index patterns that are about 3k characters long

mattkime replied:
Also seems like this functionality could be part of the data views service

stratoula replied:
This is a good idea!

mattkime replied:
I'm currently uncertain as to whether this makes sense as part of the data views api. We need to determine whether we're solving problems for just discover / dashboards /visualizations or all of kibana.

mattkime replied:
The id needs to use a hash and the type should be set to an 'esql' constant

kertal replied:
ah, that's why you were adding just the type, didn't see it, however, let's go with the string, and do the type in a follow up, with also taking care of the new type on data view level

mattkime commented:
I left some feedback since this PR is still being discussed. Its possible that this work will be picked back up.

stratoula commented:
/ci

mattkime commented on src/plugins/data/server/search/strategies/ese_search/ese_search_strategy.ts:
```diff
@@ -170,7 +171,7 @@ export const enhancedEsSearchStrategyProvider = (
     search: (request, options: IAsyncSearchOptions, deps) => {
       logger.debug(`search ${JSON.stringify(request.params) || request.id}`);
-      if (request.indexType && request.indexType !== 'rollup') {
+      if (request.indexType && request.indexType !== 'rollup' && request.indexType !== ESQL_TYPE) {
```
@lukasolson suggested taking another approach, but it would get rid of the error if there was a bad indexType provided. I want to make sure thats what he wants before implementing it.

kertal replied:
I'd suggest to take care of the ESQL_TYPE in a follow up, because it's not a requirement in this PR, right? And it's cleaner to do it separately

mattkime commented:
@stratoula I pushed some relatively minor changes, can you verify everything is working correctly?

kertal commented on packages/kbn-esql-utils/src/utils/index.ts:
```diff
@@ -0,0 +1,38 @@
+) {
+  return await dataViewsService.create({
+    title: indexPattern,
+    id: await sha256(indexPattern),
```
thx a lot for taking care of the crypto (💰 ) thingy 👍 
I think adding a scope/salt `esql-${indexPattern}` like before is something we should do, just to be save when any other consumer in Kibana having the same idea using a encrypted indexPattern for the id

stratoula commented:
/ci

stratoula commented:
@mattkime thanx a lot for your contribution!

@kertal I addressed your comments and removed the esql type from this PR to keep it clean. I agree we can do it in a separate PR!

stratoula commented:
/ci

stratoula commented:
/ci

stratoula commented:
This jest test failure seems irrelevant with the changes made in this PR. This PR was green before we close it and I don't think that any of Matt changes justify the failure.
I assume that this is a flaky test or something goes wrong in main.

mattkime commented:
I guess I don't understand the removal of the data view type. Do we want to be able to use one of these data views with searchsource? Seems like that would always be a bad code path and throwing an error would be the most immediate way of preventing this.

stratoula commented:
@mattkime Matthias suggested that you are going to do it in a follow up PR, this is why I removed that! https://github.com/elastic/kibana/pull/174736#discussion_r1472352474

stratoula commented:
/ci

kertal commented:
> I guess I don't understand the removal of the data view type. Do we want to be able to use one of these data views with searchsource? Seems like that would always be a bad code path and throwing an error would be the most immediate way of preventing this.

> @mattkime Matthias suggested that you are going to do it in a follow up PR, this is why I removed that! https://github.com/elastic/kibana/pull/174736#discussion_r1472352474

yes, when this is finished, let's aim to bring the parts of https://github.com/elastic/kibana/pull/172622 over the finishing line, that introduce a new type. this isn't required to bring this over the finishing line.

stratoula commented:
/ci

stratoula commented:
/ci

kertal commented:
> I guess I don't understand the removal of the data view type. Do we want to be able to use one of these data views with searchsource? Seems like that would always be a bad code path and throwing an error would be the most immediate way of preventing this.

I just wanted to add here, I was originally wrote of `ESQL_TYPE`, which would be bad, if the `type` was still set when `dataViews.create` was executed, in this case it would be a bad code path, I do agree. this has go hand in hand of course, and it was also done that way. So it should be fine and the next step, after this was merged should be to take care of the introduction of and consumption of `ESQL_TYPE` like done in https://github.com/elastic/kibana/pull/172622 , this might also mean to introduce a new data view package to omit running into circular dependency fun when using a constant from plugin and packages, something @mattkime and I were briefly talking about today.
